### PR TITLE
Resolver auto-scaling

### PR DESCRIPTION
**Pull request by davidf146**

### CODE REVIEW

File: api/v1/runtime_types.go [hunk 1]
```diff
@@ -29,6 +28,8 @@ type CBContainersRuntimeResolverSpec struct {
 	Affinity *coreV1.Affinity `json:"affinity,omitempty"`
 	// +kubebuilder:default:="info"
 	LogLevel string `json:"logLevel,omitempty"`
+	// +kubebuilder:default:=5
+	NodesToReplicasRatio *int32 `json:"nodesToReplicasRatio,omitempty"`
```
Comment (BenRub) [inline api/v1/runtime_types.go:32]: make as int

File: cbcontainers/state/components/runtime_resolver_deployment.go [hunk 2]
```diff
@@ -217,3 +232,27 @@ func (obj *ResolverDeploymentK8sObject) mutateVolumesMounts(container *coreV1.Co
...
+		return nil, fmt.Errorf("error creating Kubernetes client: %v", err)
+	}
+
+	// Get the list of nodes in the cluster
+	nodes, err := clientSet.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting list of nodes: %v", err)
+	}
+
+	nodesCount := int32(math.Ceil(float64(len(nodes.Items)) / float64(*nodesToReplicasRatio)))
```
Comment (BenRub) [inline cbcontainers/state/components/runtime_resolver_deployment.go:255]: if you don't change to int -> check for null value

File: cbcontainers/state/components/runtime_resolver_deployment.go [hunk 3]
```diff
@@ -217,3 +232,27 @@ func (obj *ResolverDeploymentK8sObject) mutateVolumesMounts(container *coreV1.Co
 
 	commonState.MutateVolumeMountToIncludeRootCAsVolumeMount(container)
 }
+
+func (obj *ResolverDeploymentK8sObject) getDynamicReplicasCount(nodesToReplicasRatio *int32) (*int32, error) {
+	// Get the in-cluster config
+	config, err := rest.InClusterConfig()
```
Comment (BenRub) [inline cbcontainers/state/components/runtime_resolver_deployment.go:238]: We already do the nodes list call on main.go
in order to get the k8s version
watch how we pass it to the function `NewStateApplier`
you pass another variable of NodesAmount or something (main.go:122)
and after that pass this value to the `NewResolverDeploymentK8sObject` function

File: config/crd/bases/operator.containers.carbonblack.io_cbcontainersagents.yaml [hunk 4]
```diff
@@ -1,9 +1,10 @@
+
 ---
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.9.0
+    controller-gen.kubebuilder.io/version: v0.4.1
```
Comment (BenRub) [inline config/crd/bases/operator.containers.carbonblack.io_cbcontainersagents.yaml:7]: please delete your lcoal
`bin/controller-gen` & /bin_old/controller-gen` files before running the make commands.
you can remove the all bin folder

File: controllers/runtime_components_defaults.go [hunk 5]
```diff
@@ -43,9 +45,9 @@ func (r *CBContainersAgentController) setRuntimeResolverDefaults(runtimeResolver
 		runtimeResolver.Env = make(map[string]string)
 	}
 
-	if runtimeResolver.ReplicasCount == nil {
-		defaultReplicaCount := int32(1)
-		runtimeResolver.ReplicasCount = &defaultReplicaCount
+	if runtimeResolver.NodesToReplicasRatio == nil {
```
Comment (BenRub) [inline controllers/runtime_components_defaults.go:48]: If you change to regular int, check that it is equal to zero

**Review by BenRub — CHANGES_REQUESTED** [review]

File: cbcontainers/state/components/runtime_resolver_deployment.go (hunk 3 above)
Comment (davidf146) [reply cbcontainers/state/components/runtime_resolver_deployment.go:238]: This needs to be called in the reconcile loop to get the latest updated nodes count.
I will propagate the client to `ResolverDeploymentK8sObject`

**Review by BenRub — APPROVED** [review]

**Pull request merged**